### PR TITLE
Enhanced Provisioner Extension loader

### PR DESCRIPTION
**Pull request by chtyim**
- Allows loading extensions from the system classloader with guice injection
- Derive provisioner json config from the classloader
- Supports loading config json from within the jar

### CODE REVIEW

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/provision/DefaultProvisionerConfigProvider.java [hunk 1]
```diff
@@ -47,55 +47,73 @@
...
+    if (classURL == null) {
+      LOG.warn("Unable to find provisioner class {} from its own classloader", provisionerClass);
+      return null;
+    }
+
+    // Find the json config file with name [provisioner_name].json under the provisioner ext directory.
+    // The ext directory can be found by the classloader of the provisioner.
+    // See AbstractExtensionLoader class for the directory structure
+    String provisionerName = provisioner.getSpec().getName();
+    URL classPathURL = ClassLoaders.getClassPathURL(provisionerClass.getName(), classURL);
```
Comment (tivv) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/provision/DefaultProvisionerConfigProvider.java]: Can you just use `ClassLoaders.getClassPathURL(Class<?> cls)` as it does same thing as two-step above?

File: cdap-app-fabric/src/main/resources/META-INF/services/io.cdap.cdap.runtime.spi.provisioner.Provisioner [hunk 2]
```diff
@@ -14,4 +14,5 @@
 # the License.
 #
 
-io.cdap.cdap.internal.tethering.runtime.spi.provisioner.TetheringProvisioner
+io.cdap.cdap.internal.provision.NativeProvisioner
+io.cdap.cdap.internal.tethering.runtime.spi.provisioner.TetheringProvisioner
```
Comment (tivv) [inline cdap-app-fabric/src/main/resources/META-INF/services/io.cdap.cdap.runtime.spi.provisioner.Provisioner]: Add newline

**Review by tivv — APPROVED** [review]
Minor comments, otherwise LGTM

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/provision/DefaultProvisionerConfigProvider.java (hunk 1 above)
Comment (chtyim) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/provision/DefaultProvisionerConfigProvider.java]: Done.

File: cdap-app-fabric/src/main/resources/META-INF/services/io.cdap.cdap.runtime.spi.provisioner.Provisioner (hunk 2 above)
Comment (chtyim) [reply cdap-app-fabric/src/main/resources/META-INF/services/io.cdap.cdap.runtime.spi.provisioner.Provisioner]: Done.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/provision/DefaultProvisionerConfigProvider.java [hunk 3]
```diff
@@ -47,55 +47,70 @@
...
+    }
+
+    // If we cannot find the config file in the filesystem, try to look for the json config from the classloader itself.
+    // This is the case when the .json is packaged inside the jar (e.g. system provisioners).
+    URL resource = provisionerClass.getClassLoader().getResource(provisionerName + ".json");
+    if (resource != null) {
+      return new InputStreamReader(resource.openStream(), StandardCharsets.UTF_8);
+    }
+
+    LOG.debug("Unable to find JSON config file {} for provisioner {}", configFilePath, provisionerName);
```
Comment (nliu132) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/provision/DefaultProvisionerConfigProvider.java:113]: native provisioner config will always be unfindable - do we want to skip trying to load its config?

Comment (chtyim) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/provision/DefaultProvisionerConfigProvider.java:113]: It only showed once as a debug log, I think it's ok not to special handle the native profile in this class.

**Review by nliu132 — APPROVED** [review]

**Pull request merged**